File: dotnet/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/PowerfulWeaponDecorator.cs
namespace DecoratorDesignPattern.Example1;

public class PowerfulWeaponDecorator : CharacterDecorator
{
    private readonly string _weaponName;
    private readonly int _attackBonus;

    public PowerfulWeaponDecorator(ICharacter character, string weaponName, int attackBonus)
        : base(character)
    {
        if (attackBonus <= 0)
            throw new ArgumentException("Saldırı bonusu pozitif olmalıdır.");

        _weaponName = weaponName ?? "Güçlü Silah";
        _attackBonus = attackBonus;
    }

    public override int GetAttackPower()
    {
        return base.GetAttackPower() + _attackBonus;
    }

    public override string GetDescription()
    {
        return $"{base.GetDescription()} + {_weaponName} (+{_attackBonus} Saldırı)";
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/src/PowerfulWeaponDecorator.ts
import { ICharacter } from './ICharacter';
import { CharacterDecorator } from './CharacterDecorator';

export class PowerfulWeaponDecorator extends CharacterDecorator {
  constructor(
    character: ICharacter,
    private readonly weaponName: string,
    private readonly attackBonus: number,
  ) {
    super(character);
    if (attackBonus <= 0) throw new Error('Saldırı bonusu pozitif olmalıdır.');
  }

  getAttackPower(): number { return super.getAttackPower() + this.attackBonus; }
  getDescription(): string { return `${super.getDescription()} + ${this.weaponName} (+${this.attackBonus} Saldırı)`; }
}
